Typescript errors?

Is it because of my Typescript version, 4.5.4...? Or perhaps my general naivety....?

OS: Windows 10
NPM:  7.24.1
Node: v16.10.0
Typescript: 4.5.4
Pactum: 3.1.1

When my `tsconfig.js` has `"allowJs": true`, adding a JS file that just requires pactum causes the below:

```bash
node_modules/pactum/src/exports/expect.d.ts:39:1 - error TS2309: An export assignment cannot be used in a module with other exported elements.
39 export = expect;
   ~~~~~~~~~~~~~~~~

node_modules/pactum/src/exports/reporter.d.ts:1:51 - error TS2440: Import declaration conflicts with local declaration of 'Interaction'.

1 import { InteractionRequest, InteractionResponse, Interaction } from './mock';
                                                    ~~~~~~~~~~~

node_modules/pactum/src/index.d.ts:5:25 - error TS2498: Module '"L:/src/foaf/node_modules/pactum/src/exports/expect"' uses 'export =' and cannot be used with 'export *'.

5 export * as expect from './exports/expect';
                          ~~~~~~~~~~~~~~~~~~

node_modules/pactum/src/index.d.ts:7:27 - error TS2307: Cannot find module './exports/matcher' or its corresponding type declarations.        

7 export * as matchers from './exports/matcher';
                            ~~~~~~~~~~~~~~~~~~~

node_modules/pactum/src/index.d.ts:15:25 - error TS2307: Cannot find module './exports/logger' or its corresponding type declarations.        

15 export * as logger from './exports/logger';
                           ~~~~~~~~~~~~~~~~~~

node_modules/pactum/src/models/Spec.d.ts:201:63 - error TS2314: Generic type 'Array<T>' requires 1 type argument(s).

201   withMultiPartFormData(key: string, value: string | Buffer | Array | ArrayBuffer, options?: FormData.AppendOptions): Spec;
                                                                  ~~~~~

node_modules/pactum/src/models/Spec.d.ts:398:19 - error TS2304: Cannot find name 'T'.

398   toss(): Promise<T>;
                      ~

Found 7 errors.
```

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "target": "ES2017",
    "module": "commonjs",
    "allowJs": true,
    "sourceMap": true,
    "outDir": "dist",
    "strict": true,
    "lib": [
      "esnext",
    ],
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "typeRoots": [
      "jest-fetch-mock",
      "./node_modules/@types",
      "./src/types"
    ],
    "baseUrl": ".",
    "paths": {
      "testlib/*": [
        "./test/lib/*"
      ],
      "src/*": [
        "./src/*"
      ]
    }
  }
}
```

Setting `"allowJs": false` solves the problem.

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "pactum",
-  "version": "3.1.3",
+  "version": "3.1.4",
   "description": "REST API Testing Tool for all levels in a Test Pyramid",
   "main": "./src/index.js",
   "types": "./src/index.d.ts",
@@ -60,11 +60,11 @@
   "author": "Anudeep <asa.anudeep4@gmail.com>",
   "license": "MIT",
   "dependencies": {
-    "@exodus/schemasafe": "^1.0.0-rc.4",
+    "@exodus/schemasafe": "^1.0.0-rc.6",
     "deep-override": "^1.0.2",
     "form-data-lite": "^1.0.2",
     "json-query": "^2.2.2",
-    "klona": "^2.0.4",
+    "klona": "^2.0.5",
     "lightcookie": "^1.0.25",
     "openapi-fuzzer-core": "^1.0.6",
     "pactum-matchers": "^1.1.0",
@@ -73,8 +73,8 @@
     "polka": "^0.5.2"
   },
   "devDependencies": {
-    "chai": "^4.3.4",
-    "mocha": "^9.1.2",
+    "chai": "^4.3.6",
+    "mocha": "^9.2.0",
     "nyc": "^15.1.0",
     "sinon": "^10.0.0"
   },

--- a/src/exports/expect.d.ts
+++ b/src/exports/expect.d.ts
@@ -1,4 +1,4 @@
-export interface Have {
+interface Have {
   status(code: number): void;
   header(key: string, value: any): void;
   headerContains(key: string, value: any): void;
@@ -25,11 +25,11 @@ export interface Have {
   _(handler: string, data: any): Promise<void>;
 }
 
-export interface To {
+interface To {
   have: Have;
 }
 
-export interface Expect {
+interface Expect {
   to: To;
   should: To;
 }

--- a/src/exports/mock.d.ts
+++ b/src/exports/mock.d.ts
@@ -51,6 +51,7 @@ export interface Interaction {
   flow?: string;
   strict?: boolean;
   background?: boolean;
+  clone?: boolean;
   request: InteractionRequest;
   response: InteractionResponse;
   expects?: InteractionExpectations;

--- a/src/exports/reporter.d.ts
+++ b/src/exports/reporter.d.ts
@@ -1,4 +1,4 @@
-import { InteractionRequest, InteractionResponse, Interaction } from './mock';
+import { InteractionRequest, Interaction } from './mock';
 
 export interface SpecRequest {
   method: string;
@@ -19,9 +19,7 @@ export interface InteractionCall {
   exercisedAt: string;
 }
 
-export interface Interaction {
-  request: InteractionRequest;
-  response: InteractionResponse;
+export interface ExercisedInteraction extends Interaction {
   calls: InteractionCall[];
 }
 
@@ -36,7 +34,7 @@ export interface SpecData {
   request: SpecRequest;
   response?: SpecResponse;
   recorded?: object;
-  interactions: Interaction[];
+  interactions: ExercisedInteraction[];
 }
 
 export interface Reporter {
@@ -44,7 +42,7 @@ export interface Reporter {
   afterSpec(data: SpecData): void;
   afterStep(data: object): void;
   afterTest(data: object): void;
-  afterInteraction(data: Interaction): void;
+  afterInteraction(data: ExercisedInteraction): void;
   end(): void | Promise<void>
 }
 

--- a/src/index.d.ts
+++ b/src/index.d.ts
@@ -2,9 +2,10 @@ import * as Spec from './models/Spec';
 import * as Fuzz from './models/Fuzz';
 import * as E2E from './models/E2E';
 
-export * as expect from './exports/expect';
+import expect from './exports/expect';
+export { expect };
+
 export * as handler from './exports/handler';
-export * as matchers from './exports/matcher';
 export * as mock from './exports/mock';
 export * as reporter from './exports/reporter';
 export * as request from './exports/request';
@@ -12,7 +13,6 @@ export * as response from './exports/response';
 export * as settings from './exports/settings';
 export * as stash from './exports/stash';
 export * as state from './exports/state';
-export * as logger from './exports/logger';
 
 /**
  * returns an instance of a spec

--- a/src/index.d.ts
+++ b/src/index.d.ts
@@ -46,4 +46,6 @@ export function sleep(ms: number): Promise<void>;
 
 export function clone<T>(input: T): T;
 
+export function parse(data: any): any;
+
 export namespace pactum { }

--- a/src/index.js
+++ b/src/index.js
@@ -22,6 +22,14 @@ const stash = require('./exports/stash');
 const expect = require('./exports/expect');
 const reporter = require('./exports/reporter');
 
+const processor = require('./helpers/dataProcessor');
+
+function parse(data) {
+  processor.processMaps();
+  processor.processTemplates();
+  return processor.processData(data);
+}
+
 const pactum = {
 
   mock,
@@ -59,7 +67,9 @@ const pactum = {
 
   clone(value) {
     return klona(value);
-  }
+  },
+
+  parse
 
 };
 

--- a/src/models/Interaction.model.js
+++ b/src/models/Interaction.model.js
@@ -2,6 +2,7 @@ const { setMatchingRules, getValue } = require('pactum-matchers').utils;
 const processor = require('../helpers/dataProcessor');
 const helper = require('../helpers/helper');
 const { PactumInteractionError } = require('../helpers/errors');
+const { klona: clone } = require('klona');
 const ALLOWED_REQUEST_METHODS = new Set([
   'GET',
   'POST',
@@ -188,6 +189,9 @@ class InteractionExpectations {
 class Interaction {
   constructor(raw) {
     let unprocessedResponse;
+    if (raw && raw.clone) {
+      raw = clone(raw);
+    }
     if (raw && raw.response) {
       unprocessedResponse = raw.response;
       delete raw.response;

--- a/src/models/Spec.d.ts
+++ b/src/models/Spec.d.ts
@@ -198,7 +198,7 @@ declare class Spec {
    * @see https://www.npmjs.com/package/form-data-lite
    * @see https://pactumjs.github.io/#/request-making?id=form-data
    */
-  withMultiPartFormData(key: string, value: string | Buffer | Array | ArrayBuffer, options?: FormData.AppendOptions): Spec;
+  withMultiPartFormData(key: string, value: string | Buffer | Array<any> | ArrayBuffer, options?: FormData.AppendOptions): Spec;
 
   /**
    * with http core options
@@ -395,7 +395,7 @@ declare class Spec {
    * executes the test case
    * @see https://pactumjs.github.io/#/api-testing
    */
-  toss(): Promise<T>;
+  toss(): Promise<any>;
 
   /**
    * returns chai like assertions

--- a/test/component/interactions.clone.spec.js
+++ b/test/component/interactions.clone.spec.js
@@ -1,0 +1,34 @@
+const { spec } = require('../../src/index');
+const { like } = require('pactum-matchers');
+const { expect } = require('chai');
+
+describe('Interactions - Clone', () => {
+
+  const user = {
+    name: like('guest')
+  };
+
+  it('should clone the data', async () => {
+    await spec()
+      .useInteraction({
+        clone: true,
+        request: {
+          method: 'POST',
+          path: '/api/users',
+          body: user
+        },
+        response: {
+          status: 200,
+          body: user
+        }
+      })
+      .post('http://localhost:9393/api/users')
+      .withJson({
+        name: 'guest'
+      })
+      .expectStatus(200);
+    expect(user).deep.equals({ name: { value: 'guest', pactum_type: 'LIKE' } });
+  });
+
+
+});

--- a/test/unit/parse.spec.js
+++ b/test/unit/parse.spec.js
@@ -1,0 +1,38 @@
+const { parse, stash } = require('../../src');
+const expect = require('chai').expect;
+
+const config = require('../../src/config');
+
+describe('parse', () => {
+
+  after(() => {
+    config.data.template.enabled = false;
+    config.data.template.processed = false;
+    stash.clearDataTemplates();
+    stash.clearDataMaps();
+  })
+
+  it('no data', () => {
+    const actual = parse();
+    expect(actual).to.be.undefined;
+  });
+
+  it('string', () => {
+    const actual = parse('hi mom');
+    expect(actual).to.be.equals('hi mom');
+  });
+
+  it('data map', () => {
+    stash.addDataMap({ 'User': 'guest' });
+    const actual = parse('$M{User}');
+    expect(actual).to.be.equals('guest');
+  });
+
+  it('data template', () => {
+    stash.addDataTemplate({ 'User': { name: 'guest' } });
+    const actual = parse({ '@DATA:TEMPLATE@': 'User'});
+    expect(actual).deep.equals({ name: 'guest' });
+  });
+
+});
+

--- a/test/unit/parse.spec.js
+++ b/test/unit/parse.spec.js
@@ -30,7 +30,7 @@ describe('parse', () => {
 
   it('data template', () => {
     stash.addDataTemplate({ 'User': { name: 'guest' } });
-    const actual = parse({ '@DATA:TEMPLATE@': 'User'});
+    const actual = parse({ '@DATA:TEMPLATE@': 'User' });
     expect(actual).deep.equals({ name: 'guest' });
   });
 

